plonky_block_proof_gen/src/verifier_state.rs

```diff
@@ -0,0 +1,70 @@
+//! This module defines the `VerifierState`, that contains the necessary data to
+//! handle succinct block proofs verification.
+
+use log::info;
+use plonky2::recursion::cyclic_recursion::check_cyclic_proof_verifier_data;
+
+use crate::proof_gen::ProofGenResult;
+use crate::prover_state::ProverStateBuilder;
+use crate::types::PlonkyProofIntern;
+use crate::{prover_state::ProverState, types::VerifierData};
+
+/// Plonky2 verifier state.
+///
+/// The default generation requires generating all the prover data before
+/// extracting the verifier-related data, which can take a long time and require
+/// a large amount of memory.
+pub struct VerifierState {
+    /// The verification circuit data associated to the block proof layer of the
+    /// plonky2 prover state.
+    pub state: VerifierData,
+}
+
+/// Builder for the verifier state.
+/// This is essentially the same as the [`ProverStateBuilder`], in that we need
+/// to first generate the entire prover state before extracting the verifier
+/// data.
+pub type VerifierStateBuilder = ProverStateBuilder;
+
+impl VerifierStateBuilder {
+    /// Instantiate the verifier state from the builder. Note that this is a
+    /// very expensive call!
+    pub fn build_verifier(self) -> VerifierState {
+        info!("Initializing Plonky2 aggregation verifier state (This may take a while)...");
+        let ProverState { state } = self.build(false);
+        info!("Finished initializing Plonky2 aggregation verifier state!");
+
+        VerifierState {
+            state: state.final_verifier_data(),
+        }
+    }
+}
+
+/// Extracts the verifier state from the entire prover state.
+impl From<ProverState> for VerifierState {
+    fn from(prover_state: ProverState) -> Self {
+        VerifierState {
+            state: prover_state.state.final_verifier_data(),
+        }
+    }
+}
+
+impl VerifierState {
+    /// Verifies a `block_proof`.
+    pub fn verify(&self, block_proof: &PlonkyProofIntern) -> ProofGenResult<()> {
```

Comment:
@BGluth I didn't want to bother creating a new Result / Error type just for one method, so I reused the `ProofGen` ones, but let me know if you'd rather me to create a custom pair for `ProofVerif`.

Reply:
Yeah I agree, creating a separate error seems a bit excessive. We could rename `ProofGenResult` to something that's common between `ProogGen` and `Verifier` I guess, but I don't think it's a big deal if we leave it the way it is.